pkg/controller/direct/apikeys/apikeyskey/apikeyskey_controller.go

```diff
@@ -0,0 +1,296 @@
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	key, err := a.gcp.GetKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			klog.Warningf("key was not found: %v", err)
+			return false, nil
+		}
+		return false, err
+	}
+
+	u := &v1alpha1.APIKeysKey{}
+	if err := keyMapping.Map(key, u); err != nil {
+		return false, err
+	}
+	a.actual = u
+
+	return true, nil
+}
+
+func (a *adapter) Delete(ctx context.Context) error {
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	op, err := a.gcp.DeleteKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("deleting key: %w", err)
+	}
+
+	if _, err := op.Wait(ctx); err != nil {
+		return fmt.Errorf("waiting for key deletion to complete: %w", err)
+	}
+
+	return nil
+}
+
+func (a *adapter) buildCreateRequest(ctx context.Context) (*pb.CreateKeyRequest, error) {
+	// You can configure only the `display_name`, `restrictions`, and
+	// `annotations` fields.
+	desired := &pb.Key{}
+	if err := keyMapping.Map(a.desired, desired); err != nil {
+		return nil, err
+	}
+
+	return &pb.CreateKeyRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/%s", a.projectID, a.location),
+		KeyId:  a.keyID,
+		Key:    desired,
+	}, nil
+}
+
+func (a *adapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	req, err := a.buildCreateRequest(ctx)
+	if err != nil {
+		return err
+	}
+
+	op, err := a.gcp.CreateKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating key: %w", err)
+	}
+	// TODO: Is the resourceID returned if it is dynamically created?  Maybe we should create the UUID?
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting for key creation: %w", err)
+	}
+	log := klog.FromContext(ctx)
+	log.V(2).Info("created key", "key", created)
+	// TODO: Return created object
+	return nil
+}
+
+func (a *adapter) Update(ctx context.Context) (*unstructured.Unstructured, error) {
+	// TODO: Return updated object
+
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	// TODO: I think we can do this with a helper
+	if !reflect.DeepEqual(a.desired.Spec.DisplayName, a.actual.Spec.DisplayName) {
+		updateMask.Paths = append(updateMask.Paths, "display_name")
+	}
+	if !reflect.DeepEqual(a.desired.Spec.Restrictions, a.actual.Spec.Restrictions) {
+		updateMask.Paths = append(updateMask.Paths, "restrictions")
+	}
+
+	// TODO: Annotations
+	// if !reflect.DeepEqual(a.desired.Annotations, a.actual.Annotations) {
+	// 	updateMask.Paths = append(updateMask.Paths, "annotations")
+	// }
+
+	if len(updateMask.Paths) == 0 {
+		// TODO: Log/warn/error?
+		return nil, nil
+	}
+
+	key := &pb.Key{}
+	if err := keyMapping.Map(a.desired, key); err != nil {
+		return nil, err
+	}
+
+	req := &pb.UpdateKeyRequest{
+		Key:        key,
+		UpdateMask: updateMask,
+	}
+
+	req.Key.Name = a.fullyQualifiedName()
+
+	_, err := a.gcp.UpdateKey(ctx, req)
+	if err != nil {
+		return nil, err
```

Comment:
Seems racy as there should have been a get before this but what if now the resource does not exist?